### ISSUE TITLE
How to add additional flag to kubelet for Rancher Kubernetes?

### ISSUE THREAD

**Issue by windswordsgd** [rancher/rancher#17154]
**What kind of request is this (question/bug/enhancement/feature request):**
Currently we are using Rancher 2.0 Kubernetes.  The cluster encountered one cluster level OOM issue and crashed all the applications.  We would like to add kube-reserved and pod-hard-evication to the system to prevent node OOM.    However,  we cannot find a way to update kubelet flags in Rancher GUI or within the cluster nodes.   

**Steps to reproduce (least amount of steps as possible):**
From one worker nodes,  we can inspect kubelet container but unable to make the change. 

Worker node:
docker inspect kubelet | more
[
    {
        "Id": "51c91faac9aba066cf53f792e0cbaa1f30fb4c190bd80662ff2c61c82212f524",
        "Created": "2018-12-18T04:10:54.05209824Z",
        "Path": "/opt/rke-tools/entrypoint.sh",
        "Args": [
            "kubelet",
            "--cni-bin-dir=/opt/cni/bin",
            "--resolv-conf=/etc/resolv.conf",
            "--cluster-domain=cluster.local",
            "--kubeconfig=/etc/kubernetes/ssl/kubecfg-kube-node.yaml",
            "--client-ca-file=/etc/kubernetes/ssl/kube-ca.pem",
            "--tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_EC
DHE_RSA_WITH_CHACHA20_POLY1305",
            "--cloud-config=/etc/kubernetes/cloud-config",
            "--read-only-port=0",
            "--cgroups-per-qos=True",
            "--cni-conf-dir=/etc/cni/net.d",
            "--authentication-token-webhook=true",
            "--pod-infra-container-image=rancher/pause-amd64:3.1",
            "--hostname-override=davidk8smaster1",
            "--v=2",
            "--address=0.0.0.0",
            "--cadvisor-port=0",
            "--volume-plugin-dir=/var/lib/kubelet/volumeplugins",
            "--fail-swap-on=false",
            "--root-dir=/var/lib/kubelet",
            "--register-with-taints=node-role.kubernetes.io/controlplane=true:NoSchedule",
            "--enforce-node-allocatable=",
            "--network-plugin=cni",
            "--allow-privileged=true",
            "--anonymous-auth=false",
            "--cluster-dns=10.43.0.10",
            "--cloud-provider=azure"
        ],

**Result:**

Is there a way to add new flags to kubelet?  

This is important to prevent node OOM for this system as currently the pod eviction is only 100M. Searched the document but cannot find a feasible way on a running system.

Any help is much appreciated. 

**Other details that may be helpful:**

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI):
- Installation option (single install/HA):
Single

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported):  Rancher Kubernetes on Azure VM ( Not AKS)
- Machine type (cloud/VM/metal) and specifications (CPU/memory):  VM 2C4G
- Kubernetes version (use `kubectl version`): 

Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:17:28Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.5", GitCommit:"32ac1c9073b132b8ba18aa830f46b77dcceb0723", GitTreeState:"clean", BuildDate:"2018-06-21T11:34:22Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

- Docker (use `docker info`):

Containers: 26
 Running: 10
 Paused: 0
 Stopped: 16
Images: 6
Server Version: 17.03.2-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 106
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 4ab9917febca54791c5f071a9d1f404867857fcc
runc version: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-1021-azure
Operating System: Ubuntu 16.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.853 GiB
Name: davidk8smaster1
ID: 2EYI:ACJ5:I6EV:6476:SLJI:UWIT:ID3H:SLMZ:X4T3:5TZ3:M3QN:QB7T
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
 provider=azure
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

**Comment by superseb** [comment]
Please search the docs before filing issue: https://rancher.com/docs/rancher/v2.x/en/faq/technical/#how-can-i-add-additional-arguments-binds-environment-variables-to-kubernetes-components-in-a-rancher-launched-kubernetes-cluster

**Comment by windswordsgd** [comment]
Thanks for the comments . Really appreciated.  I have set the kube-reserved flag in kubelet.  However,  it won't work if a kube-served CG group is not specified.  I checked the Rancher Kubernetes and it does not create a new CG group automatically.

Kubelet is containerized on the nodes,  are there any way or the instruction to create cgroup on each node for kube-reserved cggroup?  

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/node-allocatable.md#recommended-cgroups-setup

Specifying kube-reserved value in this flag is invalid if --kube-reserved-cgroup flag is not specified.
--kube-reserved-cgroup=<absolute path to a cgroup>
This flag helps kubelet identify the control group managing all kube components like Kubelet & container runtime that fall under the KubeReserved reservation.
Example: /kube.slice. Note that absolute paths are required and systemd naming scheme isn't supported.
